feat: Add updated readiness probe for vLLM containers

rajinator opened the pull request:
Implements model-aware health checking for Kubernetes readiness probes. Ensures pods are marked Ready only when vLLM has loaded models and API is serving requests, preventing premature traffic routing.

Changes:
- Add readiness_probe.sh script with /health and /v1/models validation
- Update Dockerfiles to include probe script in images
- Update example configurations with exec probes
- Add mandatory PORT parameter to prevent misconfigurations
- Add documentation for probe usage

Tested with simulated accelerator deployment on Kind cluster.

- Reopening from upstream, instead of fork

Closes #300

rajinator committed:
feat: Add comprehensive readiness probe for vLLM containers

Implements model-aware health checking for Kubernetes readiness probes.
Ensures pods are marked Ready only when vLLM has loaded models and
API is serving requests, preventing premature traffic routing.

Changes:
- Add readiness_probe.sh script with /health and /v1/models validation
- Update Dockerfiles to include probe script in images
- Update example configurations with exec probes
- Add mandatory PORT parameter to prevent misconfigurations
- Add documentation for probe usage

Tested with simulated accelerator deployment on Kind cluster.

Closes #300

Signed-off-by: rajinator <rajx360@gmail.com>

rajinator committed:
refactor: Pivot from exec to HTTP probes for model-aware readiness

Based on PR #330 reviewer feedback, this commit changes the readiness
checking approach from exec-based scripts to native HTTP probes using
vLLM's OpenAI-compatible API endpoints.

**Why This Change:**

Exec probes have significant drawbacks:
- Higher overhead (fork/exec on every probe)
- Incompatible with many cloud load balancers
- More complex to debug and maintain
- Requires custom scripts in container images

HTTP probes using /v1/models provide the same model-aware readiness
checking with better cloud integration and lower overhead.

**Changes:**

Container Images:
- Remove readiness_probe.sh script
- Remove script installation from all Dockerfiles (cuda, aws, xpu)

Example Configurations:
- Update simulated-accelerators values.yaml to HTTP probes
- Update wide-ep-lws manifests (decode.yaml, prefill.yaml) to HTTP probes
- Update .gitignore to exclude test artifacts

Documentation:
- Rewrite docs/readiness-probes.md to recommend HTTP probe strategy
- Document /v1/models endpoint for model-aware readiness
- Document /health endpoint for liveness checks

**Probe Strategy:**

```yaml
startupProbe:  # Model loading during initialization
  httpGet:
    path: /v1/models
    port: 8000

livenessProbe:  # Server process health
  httpGet:
    path: /health
    port: 8000

readinessProbe:  # Model ready to serve
  httpGet:
    path: /v1/models
    port: 8000
```

Tested with 3-node Kind cluster running simulated accelerators.
All decode (3 replicas) and prefill (1 replica) pods reached Ready
state with no probe failures.

Related: #300
Addresses: PR #330 feedback from @smarterclayton

Signed-off-by: rajinator <rajx360@gmail.com>

Gregory-Pereira commented:
OK so I tested it on an `--enforce-eager` example with < 100 seconds of startup time which worked perfectly:
```bash
root@interactive-pod-b8d988cd7-gksg9:/app# set -x
root@interactive-pod-b8d988cd7-gksg9:/app# alias k=kubectl
export NAMESPACE="greg"
export GATEWAY_SERVICE_NAME=$(kubectl get services | grep "inference-gateway" | awk '{print $1}')
export GATEWAY_SERVICE_ADDRESS="http://${GATEWAY_SERVICE_NAME}.${NAMESPACE}.svc.cluster.local"
# block on pod being ready status
kubectl wait pod \
    --for=condition=Ready \
    -n "${NAMESPACE}" \
    --timeout=10m \
    -l "llm-d.ai/inferenceServing=true"
# immediately dispatch models request
curl "${GATEWAY_SERVICE_ADDRESS}/v1/models" | jq
MODEL_NAME="$(curl "${GATEWAY_SERVICE_ADDRESS}/v1/models" | grep -o '"id":"[^"]*"' | \
                head -n 1 | \
                cut -d '"' -f 4)"
curl -s ${GATEWAY_SERVICE_ADDRESS}/v1/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "'"${MODEL_NAME}"'",
        "prompt": "Lets use this example curl request to trace the logs of the routing sidecar, and understand how it works!",
        "max_tokens": 50
    }' | jq
+ alias k=kubectl
+ export NAMESPACE=greg
+ NAMESPACE=greg
++ kubectl get services
++ grep inference-gateway
++ awk '{print $1}'
+ export GATEWAY_SERVICE_NAME=infra-inference-scheduling-inference-gateway-istio
+ GATEWAY_SERVICE_NAME=infra-inference-scheduling-inference-gateway-istio
+ export GATEWAY_SERVICE_ADDRESS=http://infra-inference-scheduling-inference-gateway-istio.greg.svc.cluster.local
+ GATEWAY_SERVICE_ADDRESS=http://infra-inference-scheduling-inference-gateway-istio.greg.svc.cluster.local
+ kubectl wait pod --for=condition=Ready -n greg --timeout=10m -l llm-d.ai/inferenceServing=true
pod/ms-inference-scheduling-llm-d-modelservice-decode-7c7f455djngr8 condition met
pod/ms-inference-scheduling-llm-d-modelservice-decode-7c7f455dkscx7 condition met
+ jq
+ curl http://infra-inference-scheduling-inference-gateway-istio.greg.svc.cluster.local/v1/models
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   481    0   481    0     0  25463      0 --:--:-- --:--:-- --:--:-- 26722
{
  "data": [
    {
      "created": 1761413744,
      "id": "Qwen/Qwen3-0.6B",
      "max_model_len": 40960,
      "object": "model",
      "owned_by": "vllm",
      "parent": null,
      "permission": [
        {
          "allow_create_engine": false,
          "allow_fine_tuning": false,
          "allow_logprobs": true,
          "allow_sampling": true,
          "allow_search_indices": false,
          "allow_view": true,
          "created": 1761413744,
          "group": null,
          "id": "modelperm-210137a309bf462a91841a8db60d891b",
          "is_blocking": false,
          "object": "model_permission",
          "organization": "*"
        }
      ],
      "root": "Qwen/Qwen3-0.6B"
    }
  ],
  "object": "list"
}
++ curl http://infra-inference-scheduling-inference-gateway-istio.greg.svc.cluster.local/v1/models
++ grep -o '"id":"[^"]*"'
++ head -n 1
++ cut -d '"' -f 4
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   481    0   481    0     0  75450      0 --:--:-- --:--:-- --:--:-- 80166
+ MODEL_NAME=Qwen/Qwen3-0.6B
+ jq
+ curl -s http://infra-inference-scheduling-inference-gateway-istio.greg.svc.cluster.local/v1/completions -H 'Content-Type: application/json' -d '{
        "model": "Qwen/Qwen3-0.6B",
        "prompt": "Lets use this example curl request to trace the logs of the routing sidecar, and understand how it works!",
        "max_tokens": 50
    }'
{
  "choices": [
    {
      "finish_reason": "length",
      "index": 0,
      "logprobs": null,
      "prompt_logprobs": null,
      "prompt_token_ids": null,
      "stop_reason": null,
      "text": " We are using a **Curl** command to make a request to a specific URL, and we are using a **log** to track the response. The log is a file that is being used to record the information. The key is to make",
      "token_ids": null
    }
  ],
  "created": 1761413744,
  "id": "cmpl-78e889f0-30e8-40b0-a86d-304d61e8b5d1",
  "kv_transfer_params": null,
  "model": "Qwen/Qwen3-0.6B",
  "object": "text_completion",
  "service_tier": null,
  "system_fingerprint": null,
  "usage": {
    "completion_tokens": 50,
    "prompt_tokens": 22,
    "prompt_tokens_details": null,
    "total_tokens": 72
  }
}
```
This tested that the timing is correct - that as soon as the `wait` condition is satisfied the server is `curl`able. The last thing I need to test is for something with high startup time like our `wide-ep-lws` example to ensure the numbers are correct

Gregory-Pereira commented:
Confirmed it works on wide-ep as well (was already running there) with ~20 minute start times. Might need to revisit numbers as vLLM releases its different compilation levels `-o` tack

rajinator committed:
feat: Add HTTP readiness probes to modelservice values

Configure model-aware health checks using vLLM API endpoints:
- startupProbe: /v1/models (model loading, 30min timeout)
- livenessProbe: /health (server health)
- readinessProbe: /v1/models (traffic control)

Ports: 8200 (decode), 8000 (prefill)

Updated inference-scheduling, pd-disaggregation, and
precise-prefix-cache-aware guide values files.

Addresses PR #330 feedback.

Signed-off-by: rajinator <rajx360@gmail.com>

Gregory-Pereira reviewed (approved):
/lgtm
